samples/template-operator/operator/controllers/sample_controller.go

```diff
@@ -85,20 +109,118 @@ func (r *SampleReconciler) HandleInitialState(ctx context.Context, sampleResourc
+	sampleObjUnstructured, err := runtime.DefaultUnstructuredConverter.ToUnstructured(sampleResource)
+	if err != nil {
+		return err
+	}
+
+	manifestClient, err := r.getManifestClient(logger, sampleResource.Spec.ChartFlags)
+	if err != nil {
+		sampleResource.Status.State = v1alpha1.SampleStateError
+		return r.Client.Status().Update(ctx, sampleResource)
+	}
+
+	// Use manifest library client to install a sample chart
+	ready, err := manifestClient.Install(manifestLib.InstallInfo{
+		Ctx: ctx,
+		ChartInfo: &manifestLib.ChartInfo{
+			ChartPath:   sampleResource.Spec.ChartPath,
+			ReleaseName: sampleResource.Spec.ReleaseName,
+		},
+		RemoteInfo: custom.RemoteInfo{
+			// destination cluster rest config
+			RemoteConfig: r.Config,
+			// destination cluster rest client
+			RemoteClient: &r.Client,
+		},
+		ResourceInfo: manifestLib.ResourceInfo{
+			// base operator resource to be passed for custom checks
+			BaseResource: &unstructured.Unstructured{Object: sampleObjUnstructured},
+		},
+		CheckFn: func(context.Context, *unstructured.Unstructured, *logr.Logger, custom.RemoteInfo) (bool, error) {
+			// your custom logic here to set ready state
+			return true, nil
+		},
+		CheckReadyStates: false,
+	})
+	if err != nil {
+		sampleResource.Status.State = v1alpha1.SampleStateError
+		return r.Client.Status().Update(ctx, sampleResource)
+	}
+	if ready {
+		sampleResource.Status.State = v1alpha1.SampleStateReady
+		return r.Client.Status().Update(ctx, sampleResource)
+	}
+	return nil
+}
+
+func (r *SampleReconciler) getManifestClient(logger *logr.Logger, configString string,
+) (*manifestLib.Operations, error) {
+	config := map[string]interface{}{}
+	if err := strvals.ParseInto(configString, config); err != nil {
+		return nil, err
+	}
+	// Example: Prepare manifest library client
+	return manifestLib.NewOperations(logger, r.Config, "nginx-release-name", cli.New(),
+		map[string]map[string]interface{}{
+			// check --set flags parameter for helm
+			"set": {},
+			// comma separated values of manifest command line flags
+			"flags": config,
+		})
 }
 
-func (r *SampleReconciler) HandleDeletingState(ctx context.Context, sampleResource *v1alpha1.Sample) error {
+func (r *SampleReconciler) HandleDeletingState(ctx context.Context, sampleResource *v1alpha1.Sample,
+	logger *logr.Logger) error {
 	// TODO: deletion logic here
 
-	// Example: If Deleting state, remove Finalizers
-	sampleResource.Finalizers = nil
-	return r.Client.Update(ctx, sampleResource)
+	sampleObjUnstructured, err := runtime.DefaultUnstructuredConverter.ToUnstructured(sampleResource)
+	if err != nil {
+		return err
+	}
+
+	manifestClient, err := r.getManifestClient(logger, sampleResource.Spec.ChartFlags)
+	if err != nil {
+		sampleResource.Status.State = v1alpha1.SampleStateError
+		return r.Client.Status().Update(ctx, sampleResource)
+	}
+
+	// Use manifest library client to install a sample chart
+	readyToBeDeleted, err := manifestClient.Uninstall(manifestLib.InstallInfo{
```

Comment:
can we refactor the installInfo to be the same for processing and deleting state? a lot of boilerplate otherwise